src/main.rs

```diff
@@ -295,113 +299,95 @@ fn main() {
         std::process::exit(exitcode::OK);
     }
 
-    #[cfg(unix)]
+    // Signalling
     {
-        use future::Either;
-
-        let sigint = Signal::new(SIGINT).flatten_stream();
-        let sigterm = Signal::new(SIGTERM).flatten_stream();
-        let sigquit = Signal::new(SIGQUIT).flatten_stream();
-        let sighup = Signal::new(SIGHUP).flatten_stream();
-
-        let mut signals = sigint.select(sigterm.select(sigquit.select(sighup)));
-
-        let signal = loop {
-            let signal = future::poll_fn(|| signals.poll());
-            let crash = future::poll_fn(|| graceful_crash.poll());
-            let change = future::poll_fn(|| config_watcher.poll());
-
-            let next = signal
-                .select2(crash)
-                .select2(change)
-                .wait()
-                .map_err(|_| ())
-                .expect("None stream errors");
-
-            match next {
-                Either::A((Either::A((signal, _)), _)) => {
-                    let signal = signal.expect("Signal streams never end");
-                    if signal == SIGHUP {
-                        reload_config(&opts, &mut topology, &mut rt)
-                    } else {
-                        break signal;
-                    }
-                }
-                Either::A((Either::B((_crash, _)), _)) => break SIGINT, // Trigger graceful shutdown if a component crashed
-                Either::B((_, _)) => reload_config(&opts, &mut topology, &mut rt),
-            }
+        // Arc and Mutex are here to satisfy 'static/Send constraints imposed by block_on method.
+        let signals = std::sync::Arc::new(std::sync::Mutex::new(MultiplexStream::new()));
```

Comment:
What type is not Send? From what I can tell everything should be Send if not even Sync? I still think this is more complicated than needs to be. The only reason I am concerned is that this is a decently critical piece of code.

Reply:
From looking at what you're adding to add_as looks like `Signal` is always Send/Sync and https://docs.rs/tokio-signal/0.2.8/tokio_signal/type.IoFuture.html is always Send

Reply:
Yes, but `&mut MultiplexStream` isn't send. And `block_on` is forcing us to use stream as a future, and that future requires `&mut` of the stream, and we need to construct that future multiple times. And because `block_on` requires `'static` , we can't pass a reference, plus the `Send` requirement, and we get `Arc::<Mutex<_>>`. 

I agree that this is complicated, but I don't see an alternative.